intTests/test_fold_rewrite_proof/tupletest.saw

```diff
@@ -0,0 +1,26 @@
+enable_experimental;
+
+let use_lemmas lemmas =
+    simplify (addsimps lemmas
+              (add_prelude_eqs ["foldl_cons","foldl_nil","head_gen","tail_gen"] (cryptol_ss())));
+
+let proveit p script =
+  do {
+    print (str_concat "Proving " (show_term p));
+    time (prove_print script p);
+  };
+
+import "tupletest.cry";
+
+fnc_lemma <- proveit {{ \x y z i -> (fnc' (fnc (x, y, z) i) i).0 == x }} z3;
+
+proveit {{ foldFunctionInverse }} do {
+    unfolding [ "foldFunctionInverse"
+              , "foldFunction"
+              , "foldFunction'"
+              ];
+    goal_normalize ["fnc", "fnc'"];
+    simplify (add_prelude_eqs ["foldl_cons","foldl_nil",
```

Comment:
Did you mean to use `fnc_lemma` here with the `use_lemmas` function you created above?

Also, this script would be more convincing if the `trivial` prover was used (rather than z3).

Reply:
I agree, that it would be a better example if we could figure out a way to use `fnc_lemma`, but unfortunately it doesn't match any term in the goal. I tried for a while to modify `fnc_lemma` to actually apply in this case, but couldn't figure out a way to make it work, so I just had Z3 unfold all the calls to `fnc` and `fnc'` to prove the result. After rewriting all the `foldl`s away, the resulting term looks like

```
\x y z -> (fnc' (fnc' (... (fnc' ((fnc (fnc (... (fnc (x,y,z) 0)...) 14) 15).0, y, z) 15) ... 1) 0).0 == True
```

So `fnc_lemma` needs to somehow rewrite a term of that form, probably something of the form `fnc' ((fnc tup i).0, y, z) i` but I wasn't sure what to rewrite that to, since the final `.0` doesn't apply until all 16 `fnc'` applications happen.

Reply:
The purpose of this example for now is to demonstrate rewriting folds, and it does that. So I'm going to merge this PR now. I'd love to figure out the "right" way to write this lemma, but let's not wait on that before merging this PR.